File: 0090.子集II/solution.cpp
class Solution {
public:
vector<vector<int>> subsetsWithDup(vector<int>& nums) 
{
	vector<vector<int>> answer;
	int nums_size=nums.size();
	sort(nums.begin(),nums.end());
	answer.push_back(vector<int>({}));
	for (int i=1;i<nums_size;++i)
		get_combine(nums,nums_size,i,answer,{});
	answer.push_back(vector<int>(nums));
	return answer;
}
void get_combine(vector<int> lists,int lists_size,int k,vector<vector<int>>& answer,vector<int> ans)
{
	if (k==0)
	{
        answer.push_back(ans);
		return;
	}
    int temporary;
	for (int i=0;i<lists_size;++i)
	{
		if (i!=0 && lists.at(0)==temporary)
        {
			lists.erase(lists.begin());
			continue;
		}
        ans.push_back(lists.at(0));
        temporary=lists.at(0);
		lists.erase(lists.begin());
		get_combine(lists,lists.size(),k-1,answer,ans);
		ans.pop_back();
	}
}
};

File: 0090.子集II/solution.go
var ans [][]int
var cur []int
var sz int

func subsetsWithDup(nums []int) [][]int {
	sort.Ints(nums)
	ans = make([][]int, 0)
	cur = make([]int, 0, len(nums))
	sz = len(nums)
	dfs(nums, 0, false)
	return ans
}

func dfs(nums []int, i int, chosePre bool) {
	if i == sz {
		ans = append(ans, append([]int{}, cur...))
		return
	}
	dfs(nums, i+1, false)
	if !chosePre && i > 0 && nums[i] == nums[i-1] {
		return
	}
	cur = append(cur, nums[i])
	dfs(nums, i+1, true)
	cur = cur[:len(cur)-1]
}